### PR TITLE
feat(query): add multi-shard topk cardinality query

### PR DESCRIPTION
**Pull request by alextheimer**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**New behavior :**
Adds `topkcard` query to CLI. This returns an estimate of the top `k` children of a shard key prefix across all shards.
(The former shard-local `topkcard` command has been renamed as `topkcardlocal`.)

Invoke it as:
```
./filo-cli --host <hostname/IP> [--port ...] --command topkcard --dataset <dataset> --k <k> [--active] --shardkeyprefix <shard-key-prefix>"
```

### CODE REVIEW

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 1]
```diff
@@ -116,6 +117,8 @@ case class SeriesKeysByFilters(filters: Seq[ColumnFilter],
                                startMs: Long,
                                endMs: Long) extends MetadataQueryPlan
 
+case class MetricCardinalitiesTopK(shardKeyPrefix: Seq[String], k: Int) extends MetadataQueryPlan
```
[the remark this answers is not in the document]
Comment (alextheimer) [reply query/src/main/scala/filodb/query/LogicalPlan.scala]: Resolved offline-- will include in some later PR.

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala [hunk 2]
```diff
@@ -156,6 +254,154 @@ final case class LabelValuesExec(queryContext: QueryContext,
...
+    val rvs = source match {
+      case tsMemStore: TimeSeriesMemStore =>
+        Observable.eval {
+          val prefixIter = PrefixIterator(tsMemStore, shardKeyPrefix)
+          val topkCards = prefixIter.flatMap{ pref =>
+            tsMemStore.topKCardinality(dataset, Seq(shard), pref, k)
+          }
+          // include each into an ItemSketch
+          val sketch = new ItemsSketch[String](MetricCardTopkExec.MAX_ITEMSKETCH_MAP_SIZE)
+          topkCards.foreach(card => sketch.update(card.childName, card.timeSeriesCount.toLong))
```
[the remark this answers is not in the document]
Comment (alextheimer) [reply query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:411]: (comments added to code)

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala [hunk 3]
```diff
@@ -299,6 +395,75 @@ final case class LabelCardinalityExec(queryContext: QueryContext,
...
+               (implicit sched: Scheduler): ExecResult = {
+    source.checkReadyForQuery(dataset, shard, querySession)
+    source.acquireSharedLock(dataset, shard, querySession)
+    val rvs = source match {
+      case tsMemStore: TimeSeriesMemStore =>
+        Observable.eval {
+          val topkCards = tsMemStore.topKCardinality(dataset, Seq(shard), shardKeyPrefix, k, addInactive)
+          // Include each into an ItemSketch.
+          val sketch = new ItemsSketch[String](TopkCardExec.MAX_ITEMSKETCH_MAP_SIZE)
+          topkCards.foreach(card => sketch.update(card.childName, card.tsCount.toLong))
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:450]: Need to use `activeTsCount` instead of `tsCount`

**Review by vishramachandran — APPROVED** [review]
Looks great. We may need to iterate on this API later to get both active and total cardinalities in the result set, but that can be a different PR. Thank you for the iterations.

Approved with some comments.

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala [hunk 4]
```diff
@@ -156,6 +254,177 @@ final case class LabelValuesExec(queryContext: QueryContext,
     s" startMs=$startMs, endMs=$endMs"
 }
 
+final case object MetricCardTopkExec {
+  // TODO: tune this
+  // See here for help choosing a size:
+  //   https://datasketches.apache.org/docs/Frequency/FrequentItemsErrorTable.html
+  val MAX_ITEMSKETCH_MAP_SIZE = 128;
```
[the remark this answers is not in the document]
Comment (alextheimer) [reply query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:402]: Good catch

**Review by vishramachandran — APPROVED** [review]

**Pull request merged**